<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>What the Package Does (Title Case) • Bioc2mlr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="What the Package Does (Title Case)">
<meta property="og:description" content="More about what it does (maybe more than one line)
    Use four spaces when indenting paragraphs within the Description.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Bioc2mlr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/V1_S4_to_task_conversion.html">S4 to task conversion</a>
    </li>
    <li>
      <a href="articles/V2_feature_selection_CPO.html">Feature Selection CPO</a>
    </li>
    <li>
      <a href="articles/V3_Benchmarking_Ablation.html">V3_Benchmarking_Ablation</a>
    </li>
    <li>
      <a href="articles/V4_Bioc2mlr_on_steroids_resample.html">Bioc2mlr on steroids, resampling</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      
      <div class="hidden name"><code></code></div>

    </div>

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="bioc2mlr" class="section level1">
<h1 class="hasAnchor">
<a href="#bioc2mlr" class="anchor"></a>Bioc2mlr</h1>
<p>R package to bridge between Bioconductor’s S4 complex genomic data container, to mlr, a meta machine learning aggregator package.</p>
<p>Bioc2mlr is designed to convert Bioconductor S4 assay data containers summarizedExperiment, MultiAssayExperiment into generalized machnine learning environment.</p>
<p>Bioconductor’s S4 data containers for genomic assays are popular, well established data structures. Their data architecture facilitates the application of common analytical procedures and well established statistical methodologies to large assay data. They are extensible to encompass new emerging technologies and analytical methods. However, the S4 system enforces strict constraints on the data and these constraints raise barriers for interoperability and integration with software and packages outside of Bioconductor’s repository.</p>
<p><a href="https://mlr.mlr-org.com/">mlr</a> is a comprehensive package for machine learning. It aggregates hundreds of supervised and unsupervised models and facilitates analytics such as resampling, benchmarking, tuning, and ensemble. The <a href="https://github.com/mlr-org/mlrCPO">mlrCPO</a> package extends mlr’s pre-processing and feature engineering functionality via composable Preprocessing Operators (CPO) ‘pipelines’.</p>
<p>Bioc2mlr is a compact utility package designed to bridge between these approaches. It deploys transformations of <a href="http://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> and <a href="http://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html">MultiAssayExperiment</a> S4 data structures into mlr’s expected format. It also implements Bioconductor’s popular feature selection (filtering) methods used by limma package and others, as a CPO. The vignettes present comparisons to the <a href="https://bioconductor.org/packages/release/bioc/html/MLInterfaces.html">MLInterfaces</a> package, which aims to achieve similar goals, and presents workflows for popular publicly available genomic datasets such as <a href="http://bioconductor.org/packages/release/data/experiment/html/curatedTCGAData.html">curatedTCGAData</a>.</p>
<div id="vision" class="section level2">
<h2 class="hasAnchor">
<a href="#vision" class="anchor"></a>Vision:</h2>
<p><img src="vision.jpg" width="640"></p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># Install development version from GitHub</span></a>
<a class="sourceLine" id="cb1-2" title="2">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"drorberel/Bioc2mlr"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="co"># TBA: Install release version from CRAN</span></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co"># install.packages("Bioc2mlr")</span></a></code></pre></div>
</div>
</div>
<div id="current-implementations" class="section level1">
<h1 class="hasAnchor">
<a href="#current-implementations" class="anchor"></a>Current implementations</h1>
<p>Two Bioconductor assay container are currently implemented: <a href="http://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> for a single assay (though may have multiple sub-assays slots), and <a href="http://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html">MultiAssayExperiment</a> for multiple assays. Within the machine-learning framework, the two main steps that are adapted are the pre-processing step, followed by the (multivariate) model fitting.</p>
<p>Tools will be demonstrated for each of these 4 combinations.</p>
<table class="table">
<thead><tr class="header">
<th>S4 assay data container</th>
<th align="center">Pre-processing (TBA)</th>
<th align="right">Model (multivariate)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SummarizedExperiment (SE)</td>
<td align="center">limmaCPO</td>
<td align="right">Fun_SE_to_taskFunc</td>
</tr>
<tr class="even">
<td>MultiAssayExperiment (MAE)</td>
<td align="center">UnivCPO</td>
<td align="right">Fun_MAE_to_taskFunc</td>
</tr>
</tbody>
</table>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p><a href="https://github.com/FredHutch/Bioc2mlr/tree/master/vignettes">check vignettes:</a><br><a href="S4_to_task_conversion.html">V1_S4_to_task_conversion</a><br>
TBA V2_feature_selection_CPO<br>
TBA V3_Benchmarking_Ablation<br>
TBA V4_Bioc2mlr_on_steroids_resample<br><br></p>
</div>
<div id="case-studies" class="section level2">
<h2 class="hasAnchor">
<a href="#case-studies" class="anchor"></a>Case studies:</h2>
<div id="cavdmetamae-proof-of-concept-example" class="section level3">
<h3 class="hasAnchor">
<a href="#cavdmetamae-proof-of-concept-example" class="anchor"></a>1. CAVDmetaMAE: proof-of-concept example</h3>
<p>CAVD dataspace is an online resouce to access and analyze HIV vaccine experimental assay data. It is annotated, and accessible via either online tool, and R API DataSpaceR.</p>
<p>The CAVDmetaMAE package implement a hypothesis-free approach, to find best candidates of immune biomarkers, that are associated with experimental groups, at each study (separately), and across all studies together (meta-analysis).</p>
<p>Within each study, immune biomarkers will be analyzed by both single assays, and combinations across multiple assays.<br><a href="https://github.com/drorberel/CAVDmetaMAE" title="CAVDmetaMAE">https://github.com/drorberel/CAVDmetaMAE</a><br>
Private repo. Access permission by request.</p>
<p><br><br></p>
</div>
<div id="multi-assay-customized-feature-selection-for-jdrf-data-under-review" class="section level3">
<h3 class="hasAnchor">
<a href="#multi-assay-customized-feature-selection-for-jdrf-data-under-review" class="anchor"></a>2. Multi-assay customized feature selection for JDRF data (under review)</h3>
<div id="data-curation" class="section level4">
<h4 class="hasAnchor">
<a href="#data-curation" class="anchor"></a>Data curation</h4>
<p>A. NCBI/GEO -&gt; SEs -&gt; MAE -&gt; task (DataPackageR)</p>
</div>
<div id="papers-reproducible-results" class="section level4">
<h4 class="hasAnchor">
<a href="#papers-reproducible-results" class="anchor"></a>Paper’s reproducible results</h4>
<p>B. Biomarker discovery:<br>
B.1 Feature selection: Fun_lrn_univ_only_makePrep_MaG, Fun_lrn_univ_Clusters_All_makePrep_MaG<br>
B.2 Sensitivity analysis</p>
</div>
<div id="customized-multi-assay-feature-selection" class="section level4">
<h4 class="hasAnchor">
<a href="#customized-multi-assay-feature-selection" class="anchor"></a>Customized Multi-assay feature selection</h4>
<p>C. (TBA) UnivCPO, UnivClustCPO (refactoring the above makePreprocWrapper() <br><br></p>
</div>
</div>
<div id="annotated-public-datasets-tba" class="section level3">
<h3 class="hasAnchor">
<a href="#annotated-public-datasets-tba" class="anchor"></a>3. Annotated public datasets (TBA)</h3>
<p>TCGA <a href="http://bioconductor.org/packages/release/data/experiment/html/curatedTCGAData.html">curatedTCGAData</a> Microbiome <a href="http://bioconductor.org/packages/release/data/experiment/vignettes/curatedMetagenomicData/inst/doc/curatedMetagenomicData.html">curatedMetagenomicData</a> <br><br></p>
</div>
<div id="customized-multi-assay-cpos-composable-pipelines-tba" class="section level3">
<h3 class="hasAnchor">
<a href="#customized-multi-assay-cpos-composable-pipelines-tba" class="anchor"></a>4. Customized multi-assay CPOs / composable pipelines (TBA)</h3>
<p>Omicade4CPO <a href="http://bioconductor.org/packages/release/bioc/html/omicade4.html">Omicade4</a><br>
mixomicsCPO <a href="https://cran.r-project.org/web/packages/mixOmics/index.html">mixomics</a><br><br><br></p>
</div>
<div id="analysis-workflows-tba" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-workflows-tba" class="anchor"></a>5. Analysis Workflows: (TBA)</h3>
<p>Utilize MAE to collapse genes to sets/modules/pathways<br>
Fortified: pheatmaps, ggfortify</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>What license is it under?</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>The package maintainer <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
